src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -216,6 +216,141 @@ OMapInnerNode::rm_key(omap_context_t oc, const std::string &key)
   });
 }
 
+OMapInnerNode::rm_key_range_ret
+OMapInnerNode::rm_key_range(omap_context_t oc, const std::string &first, const std::string &last)
+{
+  LOG_PREFIX(OMapInnerNode::rm_key_range);
+  DEBUGT("remove keys: {}--{}, *this: {}", oc.t, first, last, *this);
+  auto child_fiter = first == ""? iter_cbegin() : get_containing_child(first);
+  auto child_liter = last == ""? (iter_cend()-1) : get_containing_child(last);
+  assert(child_fiter != iter_cend());
+
+  return seastar::do_with(
+    child_fiter,
+    child_liter,
+    iter_t(child_fiter),
+    mutation_result_t(mutation_status_t::SUCCESS,
+      std::nullopt, std::nullopt),
+    std::map<uint16_t, OMapNodeRef>{},
+    [this, oc, &first, &last] (
+      auto &fiter,
+      auto &liter,
+      auto &iter,
+      auto &ret,
+      auto &m_entries)
+    {
+      return trans_intr::repeat(
+        [&, oc, this]() -> rm_key_range_iertr::future<seastar::stop_iteration>
+      {
+        if (iter > liter) {
+          return rm_key_range_iertr::make_ready_future<seastar::stop_iteration>(
+            seastar::stop_iteration::yes);
+        }
+        return get_child_node(oc, iter)
+          .si_then([&, oc, this](auto &&extent) {
+          return seastar::do_with(
+            iter == fiter ? first : std::string{},
+            iter == liter ? last : std::string{},
+            [&, extent = std::move(extent)] (auto &first, auto &last) {
+            return extent->rm_key_range(oc, first, last)
+              .si_then([&, oc, this]  (auto mresult) {
+                ret.status = mresult.status;
+                if (mresult.status == mutation_status_t::NEED_MERGE) {
+                  m_entries.emplace(iter->get_offset(), *(mresult.need_merge));
+                }
+                iter++;
+                switch (mresult.status) {
+                  case mutation_status_t::SUCCESS:
+                  case mutation_status_t::NEED_MERGE:
+                    return rm_key_range_iertr::make_ready_future<seastar::stop_iteration>(
+                          seastar::stop_iteration::no);
+                  case mutation_status_t::FAIL:
+                    return rm_key_range_iertr::make_ready_future<seastar::stop_iteration>(
+                      seastar::stop_iteration::yes);
+                  default: {
+                    assert("not defined status!");
+                    return rm_key_range_iertr::make_ready_future<seastar::stop_iteration>(
+                      seastar::stop_iteration::yes);
+                  }
+                }
+            });
+          });
+        });
+      }).si_then([&, oc, this] {
```

Comment:
Expand the default `&`, so it's easier to audit the lifecycle of each captured variable.